### PR TITLE
OHRM5X-1974/1975 : Implement punchIn/out screens api

### PR DESCRIPTION
**Pull request by ShehanOHRM**

### CODE REVIEW

File: src/screens/time/Punch.tsx [hunk 1]
```diff
@@ -93,34 +101,69 @@ class Punch extends React.Component<PunchProps, PunchState> {
...
-      prevProps.punchStatus?.dateTimeEditable !==
-        this.props.punchStatus?.dateTimeEditable
-    ) {
+    if (prevProps.currentRoute !== this.props.currentRoute) {
       if (
-        this.props.currentRoute === PUNCH &&
-        this.timeInterval === null &&
-        this.props.punchStatus?.dateTimeEditable === false
+        this.props.myInfo?.user.userRole === USER_ROLE_ADMIN ||
+        this.props.myInfo?.user.isSupervisor === true
```
Comment (RajithaKumara) [inline src/screens/time/Punch.tsx:107]: this we have to remove

File: src/screens/time/Punch.tsx [hunk 2]
```diff
@@ -93,34 +101,69 @@ class Punch extends React.Component<PunchProps, PunchState> {
...
+          ) {
+            this.timeInterval = setInterval(this.onAutoReload, 30000);
+          } else {
+            clearInterval(this.timeInterval);
+            this.timeInterval = null;
+          }
+        }
       }
     }
+
```
Comment (RajithaKumara) [inline src/screens/time/Punch.tsx:143]:
```jsx
const prevPunchCurrentDateTime = prevProps.punchCurrentDate + ':' + prevProps.punchCurrentTime;
const punchCurrentDateTime = this.props.punchCurrentDate + ':' + this.props.punchCurrentTime;

if (prevPunchCurrentDateTime !== punchCurrentDateTime &&
      this.props.punchStatus?.state.name === PUNCHED_IN
    )
```

File: src/screens/time/Punch.tsx [hunk 3]
```diff
@@ -142,14 +186,21 @@ class Punch extends React.Component<PunchProps, PunchState> {
 
   onRefresh = () => {
     this.props.fetchPunchStatus();
+    this.props.fetchAttendanceConfigs();
+    const date = getDateFormatFromDateObject(new Date());
+    const time = getTimeFormatFromDateObject(new Date());
+    this.props.changePunchCurrentDateTime(date, time);
```
Comment (RajithaKumara) [inline src/screens/time/Punch.tsx:192]: *

File: src/screens/time/Punch.tsx [hunk 4]
```diff
@@ -187,22 +238,24 @@ class Punch extends React.Component<PunchProps, PunchState> {
...
 
   onPressPunchButton = () => {
-    const {punchCurrentDateTime, savedNote} = this.props;
-    if (punchCurrentDateTime !== undefined) {
+    const {punchCurrentDate, punchCurrentTime, savedNote} = this.props;
+    if (punchCurrentDate !== undefined) {
       const punchRequest: PunchRequest = {
         timezoneOffset: getCurrentTimeZoneOffset(),
-        note: savedNote ? savedNote : undefined,
-        datetime: getDateSaveFormatFromDateObject(punchCurrentDateTime),
```
Comment (RajithaKumara) [inline src/screens/time/Punch.tsx:195]: I think we can remove `getDateSaveFormatFromDateObject` method if not using anymore

File: src/screens/time/Punch.tsx [hunk 5]
```diff
@@ -187,22 +238,24 @@ class Punch extends React.Component<PunchProps, PunchState> {
...
         this.props.savePunchOutRequest({
           ...punchRequest,
         });
-      } else if (this.props.punchStatus?.punchState === PUNCHED_OUT) {
+      } else if (this.props.punchStatus?.state.name === PUNCHED_OUT) {
         this.props.savePunchInRequest({
           ...punchRequest,
         });
-      } else if (this.props.punchStatus?.punchState === INITIAL) {
+      } else if (this.props.punchStatus?.state.name === INITIAL) {
```
Comment (RajithaKumara) [inline src/screens/time/Punch.tsx:258]: Not sure this state still work with 5.x APIs

File: src/screens/time/Punch.tsx [hunk 6]
```diff
@@ -410,30 +477,50 @@ class Punch extends React.Component<PunchProps, PunchState> {
...
-                            {punchStatus?.punchState === PUNCHED_IN ? (
+                            {punchStatus?.state.name === PUNCHED_IN ? (
                               <Text>{'Punched In at' + ' : '}</Text>
                             ) : null}
                           </View>
                         </View>
                         <View style={[styles.flexFour]}>
                           <Text>
-                            {punchStatus?.punchTime
+                            {punchStatus?.state.name === PUNCHED_IN
```
Comment (RajithaKumara) [inline src/screens/time/Punch.tsx:490]: *

File: src/screens/time/Punch.tsx [hunk 7]
```diff
@@ -233,9 +286,21 @@ class Punch extends React.Component<PunchProps, PunchState> {
...
+      punchCurrentTime,
+      savedNote,
+      punchAttendanceConfig,
+      myInfo,
+    } = this.props;
     const {note} = this.state;
-    const editable = punchStatus?.dateTimeEditable;
+    const editable =
+      myInfo?.user.userRole === USER_ROLE_ADMIN ||
+      myInfo?.user.isSupervisor === true
```
Comment (RajithaKumara) [inline src/screens/time/Punch.tsx:301]: neet to check this

File: src/screens/time/PunchRequestSuccess.tsx [hunk 8]
```diff
@@ -37,8 +37,8 @@ import {
   savePunchInRequest,
   savePunchOutRequest,
 } from 'store/time/punch/actions';
-import {PUNCH_IN, PUNCH_OUT, PunchAction} from 'store/time/punch/types';
-import {PunchRequestSuccessParam} from 'screens/time/navigators/index';
+import {PUNCHED_OUT} from 'store/time/punch/types';
+import {PunchRequestSuccessParam} from './navigators';
```
Comment (RajithaKumara) [inline src/screens/time/PunchRequestSuccess.tsx:41]:
```suggestion
import {PunchRequestSuccessParam} from 'screens/time/navigators/index';
```
:thinking:

File: src/services/endpoints.ts [hunk 9]
```diff
@@ -21,9 +21,9 @@
 import {EndpointError} from 'services/errors/endpoints';
 
 // attendance
-export const API_ENDPOINT_PUNCH_STATUS = '/api/v1/attendance/punch-status';
-export const API_ENDPOINT_PUNCH_IN_REQUEST = '/api/v1/attendance/punch-in';
-export const API_ENDPOINT_PUNCH_OUT_REQUEST = '/api/v1/attendance/punch-out';
+export const API_ENDPOINT_PUNCH_STATUS = '/api/v2/attendance/records/latest';
+export const API_ENDPOINT_PUNCH_IN_OUT_REQUEST = '/api/v2/attendance/records';
+export const API_ENDPOINT_ATTENDANCE_CONFIG = '/api/v2/attendance/configs';
```
Comment (RajithaKumara) [inline src/services/endpoints.ts:26]: this may be a admin users endpoint

File: src/store/time/punch/reducer.ts [hunk 10]
```diff
@@ -40,13 +42,15 @@ const attendanceReducer = (
...
+      if (action.payload?.state.name === PUNCHED_IN) {
+        return {
+          ...state,
+          punchStatus: action.payload,
+        };
+      } else if (action.payload?.state.name === PUNCHED_OUT) {
         return {
           ...state,
           punchStatus: action.payload,
-          punchCurrentDateTime: getUTCDateObjectFromSaveFormat(
```
Comment (RajithaKumara) [inline src/store/time/punch/reducer.ts:47]: remove `getUTCDateObjectFromSaveFormat` method if not using anymore

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/screens/time/Punch.tsx (hunk 7 above)
Comment (ShehanOHRM) [reply src/screens/time/Punch.tsx:301]: remove this part

**Comment by RajithaKumara** [comment]
closing as moved to #53

**Pull request closed without merging**